source/cppexpose/source/base/Tokenizer.cpp

```diff
@@ -0,0 +1,792 @@
+
+#include <cppexpose/base/Tokenizer.h>
+
+#include <stdlib.h>
+#include <algorithm>
+#include <string>
+#include <fstream>
+
+#include <cppassist/logging/logging.h>
+#include <cppassist/string/regex.h>
+
+#include <cppexpose/base/string_helpers.h>
+
+
+static const int           MIN_INT  = int( ~(unsigned(-1)/2) );
+static const int           MAX_INT  = int(  (unsigned(-1)/2) );
+static const unsigned int  MAX_UINT = unsigned(-1);
+
+
+using namespace cppassist;
+
+
+namespace cppexpose
+{
+
+
+struct CompareStringLength
+{
+    bool operator() (const std::string & first, const std::string & second) {
+        return first.size() > second.size();
+    }
+};
+
+
+Tokenizer::Tokenizer()
+: m_options(0)
+, m_whitespace(" \t\r\n")
+, m_quotationMarks("\"")
+, m_singleCharacters("{}[],:")
+, m_begin(nullptr)
+, m_end(nullptr)
+, m_current(nullptr)
+, m_line(1)
+, m_column(1)
+, m_lastCR(false)
+{
+}
+
+Tokenizer::~Tokenizer()
+{
+}
+
+unsigned int Tokenizer::options() const
+{
+    return m_options;
+}
+
+void Tokenizer::setOptions(unsigned int options)
+{
+    m_options = options;
+}
+
+bool Tokenizer::hasOption(Tokenizer::Option option) const
+{
+    return ((m_options & option) != 0);
+}
+
+std::string Tokenizer::whitespace() const
+{
+    return m_whitespace;
+}
+
+void Tokenizer::setWhitespace(const std::string & whitespace)
+{
+    m_whitespace = whitespace;
+}
+
+std::string Tokenizer::quotationMarks() const
+{
+    return m_quotationMarks;
+}
+
+void Tokenizer::setQuotationMarks(const std::string & quotationMarks)
+{
+    m_quotationMarks = quotationMarks;
+}
+
+std::string Tokenizer::singleCharacters() const
+{
+    return m_singleCharacters;
+}
+
+void Tokenizer::setSingleCharacters(const std::string & singleCharacters)
+{
+    m_singleCharacters = singleCharacters;
+}
+
+const std::vector<std::string> & Tokenizer::standalones() const
+{
+    return m_standalones;
+}
+
+void Tokenizer::setStandalones(const std::vector<std::string> & standalones)
+{
+    m_standalones = standalones;
+
+    // Sort by string length (longest first)
+    CompareStringLength compare;
+    std::sort(m_standalones.begin(), m_standalones.end(), compare);
```

Comment:
With the use of lambdas, no explicit struct us required without any performance or space consumption impact.
```cpp
std::sort(m_standalones.begin(), m_standalones.end(), [](const std::string & first, const std::string & second) { return first.size() > second.size(); });
```